fab/core.py

```diff
@@ -104,107 +126,88 @@ def flow_alpha_2_div_nis(self, batch_size: int) -> torch.Tensor:
         loss = - torch.mean(torch.exp(2*(log_p_x - log_q_x)).detach() * log_q_x)
         return loss
 
-    def p2_over_q_alpha_2_div(self, batch_size: int) -> torch.Tensor:
+    def fab_p2_over_q_alpha_2_div_inner(self, point: Point, log_w_ais: torch.Tensor) -> \
+            torch.Tensor:
+        return - torch.mean(torch.softmax(log_w_ais, dim=-1) * point.log_q)
+
+    def fab_p2_over_q_alpha_2_div(self, batch_size: int) -> torch.Tensor:
         """Compute the FAB loss with p^2/q as the AIS target."""
         # set ais target distribution to p^2/q
-        self.set_ais_target("p^2/q")
-        if isinstance(self.annealed_importance_sampler.transition_operator, HamiltonianMonteCarlo):
```

Comment:
The `create_point` function used within the transition operator, make sure that we only have grads through `log_q(x)` and not through the `x` of the MCMC chain, which means that we do not need to block gradient here anymore.